lib/slipstream/socket_test.ex

```diff
@@ -0,0 +1,659 @@
+             )
+           ) do
+    quote do
+      return =
+        assert_receive command(%Slipstream.Commands.LeaveTopic{
+                         socket: socket,
+                         topic: unquote(topic_expr) = topic
+                       }),
+                       unquote(timeout)
+
+      send(
+        socket.socket_pid,
+        event(%Slipstream.Events.TopicLeft{topic: topic})
+      )
+
+      return
+    end
+  end
+
+  @doc """
+  Refutes that a client will request to leave a topic
+
+  The opposite of `assert_leave/3`.
+
+  ## Examples
+
+      accept_connect(MyClient)
+      assert_join "rooms:lobby", %{}, :ok
+      push(MyClient, topic, "no-don't-go", %{})
+      refute_leave ^topic, 10_000
+  """
+  @doc since: "0.2.0"
+  @spec refute_leave(topic_expr :: Macro.t(), timeout()) :: term()
+  defmacro refute_leave(
+             topic_expr,
+             timeout \\ Application.fetch_env!(
+               :ex_unit,
+               :refute_receive_timeout
+             )
+           ) do
+    quote do
+      refute_receive command(%Slipstream.Commands.LeaveTopic{
+                       topic: unquote(topic_expr)
+                     }),
+                     unquote(timeout)
+    end
+  end
+
+  @doc """
+  Asserts that the client will request to push a messsage to the server
+
+  Note that `topic_expr`, `event_expr`, `params_expr`, and `ref_expr` are all
+  pattern expressions. Prior bindings may be used with the `^` pin operator,
+  values may be underscored to ignore, and partial values may be matched (e.g.
+  `%{}` will match any map).
+
+  `ref_expr` can be provided to bind a reference for later use in `reply/3`.
+
+  ## Examples
+
+      assert_push "rooms:lobby", "msg:new", params, ref
+      reply(MyClient, ref, {:ok, %{status: "ok", received: params}})
+  """
+  @doc since: "0.2.0"
+  @spec assert_push(
+          topic_expr :: Macro.t(),
+          event_expr :: Macro.t(),
+          params_expr :: Macro.t(),
+          ref_expr :: Macro.t(),
+          timeout()
+        ) :: term()
+  defmacro assert_push(
+             topic_expr,
+             event_expr,
+             params_expr,
+             ref_expr \\ quote(do: _),
+             timeout \\ Application.fetch_env!(
+               :ex_unit,
+               :assert_receive_timeout
+             )
+           ) do
+    quote do
+      # incrementing integer? who needs it
+      unquote(ref_expr) = ref = make_ref() |> inspect()
+
+      return =
+        assert_receive gen_server_call(
+                         command(%Slipstream.Commands.PushMessage{
+                           topic: unquote(topic_expr) = topic,
+                           event: unquote(event_expr) = event,
+                           payload: unquote(params_expr) = payload
+                         }),
+                         from
+                       ),
+                       unquote(timeout)
+
+      GenServer.reply(from, ref)
+
+      return
```

Comment:
this implementation is a bit fancy because the `%Slipstream.Commands.PushMessage{}` command is sent via `GenServer.call/3` instead of `Kernel.send/2`, because the caller needs a return value from the connection (the `ref`: unique reference for the push) in order to handle replies via `c:Slipstream.handle_reply/3` or `Slipstream.await_reply/2`

the test process is the connection process (genserver) with this testing framework, so we need to fake a regular ole' GenServer call by matching on the signature `{:"$gen_call", from, message}` and replying to that `from` with `GenServer.reply/2` in order to return the ref